nabs/event.py

```diff
@@ -0,0 +1,109 @@
+from collections import defaultdict
+from functools import partial
+from threading import Event, Thread
+import math
+import time
+
+from ophyd.status import Status
+
+
+class EventBuilder:
+    def __init__(self, signals, rate, timeout=10.0, duplicates='closest'):
+        self.signals = signals
+        self.rate = rate
+        self.timeout = timeout
+        self.duplicates = duplicates
+        self.cbid = {}
+        self.start_ts = 0
+        self.name = 'event_builder'
+        self._clear_events()
+
+    def add(self, name, *, value, timestamp, **kwargs):
+        if timestamp > self.start_ts:
+            bin_ts = self._nearest_ts(timestamp)
+            delta = timestamp - bin_ts
+            event = self.bins[bin_ts]
+            event[name].append({'value': value,
+                                'timestamp': timestamp,
+                                'delta': delta})
+            if len(event) == len(self.signals):
+                self._has_evt.set()
+
+    def _nearest_ts(self, timestamp):
+        expanded = timestamp * self.rate
+        floor = math.floor(expanded) / self.rate
+        ceil = math.ceil(expanded) / self.rate
+        if timestamp - floor < ceil - timestamp:
+            return floor
+        else:
+            return ceil
+
+    def _monitor_events(self):
+        if not self.cbid:
+            self.start_ts = time.time()
+            for sig in self.signals:
+                cbid = sig.subscribe(partial(self.add, sig.name))
+                self.cbid[sig.name] = cbid
+
+    def _stop_events(self):
+        if self.cbid:
+            for sig in self.signals:
+                sig.unsubscribe(self.cbid[sig.name])
+        self.cbid = {}
+
+    def _clear_events(self):
+        self._stop_events()
+        self._has_evt = Event()
+        self.bins = defaultdict(partial(defaultdict, list))
+
+    def trigger(self):
+        self._clear_events()
+        self._monitor_events()
+        status = Status(obj=self)
+        Thread(target=self._wait_trigger, args=(status,)).start()
+        return status
+
+    def _wait_trigger(self, status):
+        success = self._has_evt.wait(timeout=self.timeout)
```

Comment:
So `trigger` will wait and then `read` will return immediately? For some reason I would have written it inverted but I can't think of a way to justify this 😄. Any precedent elsewhere, i.e `AreaDetector`?

Reply:
Second thought. This is probably the correct way.

Reply:
Yeah, this is how the synthetic devices work, so I just copied that pattern. I think this is preferred in general because you wait for triggers to complete in parallel, and then do the reads in series. Of course the distinction between `trigger` and `read` is totally lost if you only do it on one device.